### PR TITLE
Enable Paging When Retrieving NS Groups

### PR DESCRIPTION
**Pull request by neilcx**
- implement retrieving nsgroups in a paginated way
 - Add and Modify Unit Test
 - Add Integration Test

 -https://www.pivotaltracker.com/n/projects/2110693/stories/164348522

# Description

When CPI try to list NS Groups on NSX-T, it only loads for the first page. This lead to "NS Groups NOT Found" Error when the number of NS Groups is larger than page size ( default : 50 ). We should make CPI lists all pages and get a whole set of existing NS Groups instead.

## Related PR and Issues
Fixes # https://github.com/cloudfoundry/bosh-vsphere-cpi-release/issues/199

## Type of change
- [X] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
Unit Test
Integration Test

# Checklist:
- [X] My code follows the standard ruby style guide
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb [hunk 1]
```diff
@@ -633,4 +634,43 @@
...
+    end
+
+    context 'when there are two pages' do
+      let(:result_list_1) { double(Object, results: [nsgroup_1, nsgroup_2], cursor: 1) }
+      let(:result_list_2) { double(Object, results: [nsgroup_3, nsgroup_4], cursor: nil) }
+      let(:result_list_1_cursor) { "fake-result-list-1-cursor"}
+      it 'returns all NS Groups in the two pages' do
+        allow_any_instance_of(VSphereCloud::NSXTProvider).to receive(:grouping_obj_svc).and_return(grouping_obj_svc)
+        allow(grouping_obj_svc).to receive(:list_ns_groups).and_return(result_list_1)
+        allow(grouping_obj_svc).to receive(:list_ns_groups).with(anything).and_return(result_list_2)
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Can you also put a expect 

`expect(grouping_obj_svc).to receive(:list_ns_groups).with(any_args).twice`

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb [hunk 2]
```diff
@@ -633,4 +634,43 @@
...
+    let(:nsgroup_2) do
+      NSXT::NSGroup.new(:id => 'id-2', :display_name => 'test-nsgroup-2')
+    end
+    let(:nsgroup_3) do
+      NSXT::NSGroup.new(:id => 'id-3', :display_name => 'test-nsgroup-3')
+    end
+    let(:nsgroup_4) do
+      NSXT::NSGroup.new(:id => 'id-4', :display_name => 'test-nsgroup-4')
+    end
+
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Can you add a test where the result list is empty? pages == 0

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb [hunk 3]
```diff
@@ -219,7 +220,7 @@
 
     before do
       allow_any_instance_of(VSphereCloud::NSXTProvider).to receive(:grouping_obj_svc).and_return(grouping_obj_svc)
-      allow(grouping_obj_svc).to receive_message_chain(:list_ns_groups, :results).and_return([nsgroup_1, nsgroup_2])
+      allow(nsxt_provider).to receive(:retrieve_all_ns_groups_with_pagination).and_return([nsgroup_1, nsgroup_2])
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Use `expect` instead of `allow`

Expect makes unit test more water-tight by counting the number of times it is called. If your program is non-deterministic about number of calls, `allow` is fine. Otherwise, use `expect`

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb [hunk 4]
```diff
@@ -193,6 +193,7 @@
         end
 
         it 'adds simple expressions containing the logical ports to each NSGroup' do
+          allow(nsxt_provider).to receive(:retrieve_nsgroups).with(ns_groups).and_return([nsgroup_1, nsgroup_2])
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Use `expect`

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb [hunk 5]
```diff
@@ -179,7 +179,7 @@
         let(:ns_groups) { %w(other-nsgroup-1 other-nsgroup-2) }
 
         it 'should raise a NSGroupsNotFound error' do
-          expect(grouping_obj_svc).to receive_message_chain(:list_ns_groups, :results).and_return([nsgroup_1, nsgroup_2])
+          allow(nsxt_provider).to receive(:retrieve_all_ns_groups_with_pagination).and_return([nsgroup_1, nsgroup_2])
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Use `expect`

File: src/vsphere_cpi/spec/integration/nsxt_spec.rb [hunk 6]
```diff
@@ -407,6 +407,54 @@
...
+  describe 'on adding a vm to security groups' do
+    let(:vm_type) do
+      {
+        'ram' => 512,
+        'disk' => 2048,
+        'cpu' => 1,
+        'nsxt' => nsxt_spec
+      }
+    end
+    let(:nsxt_spec) do
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/integration/nsxt_spec.rb]: Just put it back up in vm_type above. it is a single line and you are not modifying it in any examples below.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb (hunk 3 above)
Comment (neilcx) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Done

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb (hunk 2 above)
Comment (neilcx) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Done

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb (hunk 1 above)
Comment (neilcx) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Done.

File: src/vsphere_cpi/spec/integration/nsxt_spec.rb [hunk 7]
```diff
@@ -407,6 +407,54 @@
...
+      {
+        'ns_groups' => @names
+      }
+    end
+
+    before do
+      @names = []
+      @groups = []
```
[the remark this answers is not in the document]
Comment (neilcx) [reply src/vsphere_cpi/spec/integration/nsxt_spec.rb]: Done.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb (hunk 4 above)
Comment (neilcx) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Done.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb (hunk 5 above)
Comment (neilcx) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb]: Done.

File: src/vsphere_cpi/spec/integration/nsxt_spec.rb (hunk 6 above)
Comment (neilcx) [reply src/vsphere_cpi/spec/integration/nsxt_spec.rb]: Done.
